ecs-init/engine/engine.go

```diff
@@ -165,10 +167,36 @@ func (e *Engine) load(image io.ReadCloser, err error) error {
 	return e.downloader.RecordCachedAgent()
 }
 
+// retryDelay returns an exponential backoff for delaying the service restart.
+// delay will look something like this:
+//     Retry 1: 2.162s
+//     Retry 2: 7.548s
+//     Retry 3: 14.776s
+//     Retry 4: 16.944s
+//     Retry 5: 34.592s
+//     Retry 6: 1m0.425s
+//     Retry 7: 1m0.456s
+//     Retry 8: 1m0.694s ...etc...
+func retryDelay(retryCount int) time.Duration {
+	minTime := 1000
+	maxTime := time.Second * 60
+
+	// (1<<n is 2 to the power n)
+	delay := time.Duration((1<<uint(retryCount))*(rand.Intn(minTime)+minTime)) * time.Millisecond
+	if delay > maxTime {
+		return maxTime + time.Duration(rand.Intn(minTime))*time.Millisecond
+	}
+	return delay
+}
+
 // StartSupervised starts the ECS Agent and ensures it stays running, except for terminal errors (indicated by an agent exit code of 5)
 func (e *Engine) StartSupervised() error {
 	agentExitCode := -1
+	retryCount := 0
 	for agentExitCode != terminalSuccessAgentExitCode && agentExitCode != terminalFailureAgentExitCode {
+		if retryCount > 0 {
+			time.Sleep(retryDelay(retryCount))
```

Comment:
can you log the delay being used here? i think it's better if customers (and even ourselves) know about how long we need to wait

Reply:
yes will do 👍